RFC.md

```diff
@@ -1,12 +1,289 @@
+as `multidim_array_index A d1, d2, d3, ... dk` instead of flattening
+the information into `gep A, d1 * n1 + d2 * n2 + ... + dk * nk`. The former
+unflattened representation is advantageous for analysis and optimisation. It also
+allows us to represent array indexing semantics of languages such as Fortran
+and Chapel, which differs from that of C based array indexing.
+	
+## Motivating the need for a multi-dimensional array indexing intrinsic
+There are primarily one of two views of array indexing involving
+multiple dimensions most languages take, which we discuss to motivate 
+the need for a multi-dimensional array index. This consideration impacts
+the kinds of analysis we can perform on the program. In Polly, we care about
+dependence analysis, so the examples here will focus on that particular problem.
 
-# The math, how to express every conceivable array indexing scheme known to man
+Let us consider an array indexing operation of the form:
+```cpp
+int ex1(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+	assert(x1 != x2);
+	assert(y1 != y2);
+	B[x1][y1] = 1;
+	printf("%d", B[x2][y2]);
+	exit(0);
+}
+```
 
+One would like to infer that since the array indices _interpreted as tuples_
+`(x1, y1)` and `(x2, y2)` do not have the same value, due to the guarding asserts
+that `x1 != x2` and `y1 != y2`. The write `B[x1][y1] = 1` can 
+in no way interfere with the value of `B[x2][y2]`. Consquently,
+we can optimise the program into:
+
+
+```cpp
+int ex1_opt(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+    // B[x1][y1] = 1 is omitted because the result 
+    // cannot be used: 
+    //  It is not used in the print and then the program exits
+	printf("%d", B[x2][y2]);
+	exit(0);
+}
+```
+
+However, alas, this is illegal, for the C language does not provide
+semantics that allow the final inference above. It is conceivable that
+`x1 != x2, y1 != y2`, but the indices do actually alias, since
+according to C semantics, the two indices alias if the _flattened
+representation of the indices alias_. Consider the parameter
+values:
+
+```
+n = m = 3
+x1 = 1, y1 = 0; B[x1][y1] = nx1+y1 = 3*1+0=3
+x2 = 0, y2 = 3; B[x2][y2] = nx2+y2 = 3*0+3=3
+```
+
+Hence, the array elements `B[x1][y1]` and `B[x2][y2]` _can alias_, and
+so the transformation proposed in `ex1_opt` is unsound in general.
+
+
+In contrast, many langagues other than C require that index
+expressions for multidimensional arrays have each component within the
+array dimension for that component. As a result, in the example above,
+the index pair `(0,3)` would be out-of-bounds. In languages with these
+semantics, one can infer that the indexing:
+
+`[x1][y1] != [x2][y2]` iff `x1 != x2 || y1 != y2`.
+
+While this particular example is not very interesting, it shows the
+spirit of the lack of expressiveness in LLVM we are trying to
+improve. There is a more realistic (and involved) example in 
+[Appendix A](#Appendix-A).
```

Comment:
I appreciate this link to Appendix A but it breaks up the flow. I'd consider moving it to just before "However, alas, this is illegal" or else to the end of this section.